src/bms/parse.rs

```diff
@@ -43,8 +40,36 @@ pub enum ParseWarning {
     PromptHandlerWarning,
 }
 
-/// type alias of core::result::Result<T, ParseWarning>
-pub(crate) type Result<T> = core::result::Result<T, ParseWarning>;
+/// type alias of core::result::Result<T, ParseWarningContent>
+pub(crate) type Result<T> = core::result::Result<T, ParseWarningContent>;
+
+/// A parse warning with position information.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct ParseWarning {
+    /// The content of the parse warning.
+    pub content: ParseWarningContent,
+    /// The row (line number) where the warning occurred.
+    pub row: usize,
+    /// The column (character position) where the warning occurred.
+    pub col: usize,
+}
+
+impl std::fmt::Display for ParseWarning {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{} at line {}, column {}",
+            self.content, self.row, self.col
+        )
+    }
+}
+
+impl std::error::Error for ParseWarning {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(&self.content)
+    }
+}
```

Comment:
```suggestion
#[derive(Debug, Clone, PartialEq, Eq, Hash, Error)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct ParseWarning {
    /// The content of the parse warning.
    #[source]
    pub content: ParseWarningContent,
    /// The row (line number) where the warning occurred.
    pub row: usize,
    /// The column (character position) where the warning occurred.
    pub col: usize,
}

impl std::fmt::Display for ParseWarning {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} at line {}, column {}",
            self.content, self.row, self.col
        )
    }
}
```
You can write this more simply with `thiserror::Error` derive macro.